Bottom: 3af064a22089ea8866f0f9d1873ff0e5c426c7dc
Top:    e959c3b0d67a3b16e5243fd69184937bf59c0b38
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-08-16 09:17:44 -0500

Add codebase generator function

The codebase generator takes information about a change, like repository,
branch, and revision, and creates a predictable string describing the code.
This is used to decide upon the builder for a change and also to tie a fork
repository back to its parent for merge requests.


---
diff --git a/docker/buildbot-master/buildbot/master.cfg b/docker/buildbot-master/buildbot/master.cfg
index aa317118..5b160a33 100644
--- a/docker/buildbot-master/buildbot/master.cfg
+++ b/docker/buildbot-master/buildbot/master.cfg
@@ -6,6 +6,7 @@
 
 import os
 from buildbot.plugins import secrets, steps, util
+from urlparse import urlparse
 
 os.environ['LONG_RUN_TIMEOUT'] = '14400'
 
@@ -140,3 +141,27 @@ c['workers'].extend(ptxdist.workers)
 c['schedulers'].extend(ptxdist.schedulers)
 c['builders'].extend(ptxdist.builders)
 c['change_source'].extend(ptxdist.change_source)
+
+
+def codebaseGenerator(chdict):
+    """ Generate a codebase description.
+
+    The codebase generator takes information about a change, like repository,
+    branch, and revision, and creates a predictable string describing the code.
+    This is used to decide upon the builder for a change and also to tie a fork
+    repository back to its parent for merge requests.
+    """
+    path = urlparse(chdict['revlink']).path.lstrip('/')
+
+    components = []
+    while path != "":
+        path, tail = os.path.split(path)
+        if tail != "":
+            components.append(tail)
+    components.reverse()
+
+    codebase = os.path.join(*components[0:2])
+    return codebase
+
+
+c['codebaseGenerator'] = codebaseGenerator
